Maps: Add new basic maps application and maps widget

bplaat opened the pull request:
I have created a basic Maps application that uses the OpenStreetMap tile service. This is my first SerenityOS userland / application so it is maybe quite bad. I first prototype the math and map rending with a [HTML Canvas website](https://gist.github.com/bplaat/b60bb09ca45770dc8eaaf7fb09b919a3). Then I just ported that to SerenityOS C++, so architecture things can be made better because it is mostly a JavaScript port.

# Demo video
https://github.com/SerenityOS/serenity/assets/19894029/a310df40-e912-47d8-96fd-0df1dd997794

*This is SerenityOS x86_64 running on a ARM64 platform so the performance looks worse then it is*

When this is merged we can add more stuff in the future like:
- Request nice Maps app icon in #artwork with a globe and a compass
- Maps Settings: select tiles maps provider
- Option to load [usermap.serenityos.org](https://usermap.serenityos.org/) data so you can watch where people live right in SerenityOS!
- Search bar with search from [OpenStreetMap API](https://wiki.openstreetmap.org/wiki/API_v0.6)
- Split of MapsWidget to LibGUI so other applications can use a map widget with an API like Leaflet.js
    - For example: In FileManager when viewing image metadata show image geolocation tag in a MapWidget
- Remember favorite places
- Geolocation: new Lib needed that does some IP location lookup (can also be used in Ladybird for the Geolocation API)

Looking forward to feedback!

AtkinsSJ commented on Userland/Applications/Maps/MapsWidget.h:
```diff
@@ -6,6 +6,7 @@
 
 #pragma once
 
+#include <AK/URL.h>
```
I don't think you're using this include

bplaat replied:
You are right

Xexxa commented:
I don't know much about OpenStreetMap tile service. But isn't relying on OpenStreetMap a SPOF? Wouldn't it be good to have a setting to change source to make sure the app works even if OpenStreetMap down? Seems to be a bunch of compatible services at https://wiki.openstreetmap.org/wiki/Raster_tile_providers

bplaat commented:
Good point, I will work on a settings window where you can choose a map tiles provider or set a custom tiles API url. I will do this in another future pull request to keep this one small and reviewable.

AtkinsSJ commented on Userland/Applications/Maps/MapsWidget.h:
```diff
@@ -0,0 +1,72 @@
+    bool m_dragging { false };
+    int m_last_mouse_x { 0 };
+    int m_last_mouse_y { 0 };
+    Vector<Tile*> m_tiles;
```
What's the reason for storing pointers here instead of the Tiles themselves?

bplaat replied:
There was no reason really, I have reworked the code so that just `Tile`s are stored

AtkinsSJ commented on Userland/Applications/Maps/MapsWidget.cpp:
```diff
@@ -0,0 +1,182 @@
+    auto request = m_request_client->start_request("GET", url, headers, {});
+
+    // Ref it here so it wont go out of scope
+    request->ref();
```
This feels sketchy. Maybe the MapsWidget could store the active requests instead?

AtkinsSJ commented on Userland/Applications/Maps/MapsWidget.cpp:
```diff
@@ -0,0 +1,182 @@
+    // Adjust map center latitude and longitude by mouse delta
+    double tile_size = TILE_SIZE * pow(2, m_zoom);
+    m_center.latitude -= (-delta_y / tile_size) * 180;
+    m_center.latitude = min(max(m_center.latitude, -MAX_LATITUDE), MAX_LATITUDE);
```
The latitude movement feels a bit off. Dragging vertically doesn't quite follow the cursor. (Dragging from the top to the bottom of the screen, the map moves about 2/3 as far as it should.) Longitude is perfect though.

bplaat replied:
You are right the math was wrong. I have reworked and simplified the math in the new code and dragging works correct now.

AtkinsSJ reviewed (changes requested):
This is cool! While playing around with it though, I had it freeze. Here's the stack trace:
![image](https://github.com/SerenityOS/serenity/assets/222642/86f54c20-7160-47c1-a542-8ebc992bfeac)

bplaat commented:
> This is cool! While playing around with it though, I had it freeze. Here's the stack trace:

I have fixed the issues you had pointed out. I can't really find where the initial freeze is coming from. Put the wait is in `LibProtocol` so that is outside the scope of this pull request. I think I just flood the RequestServer with many requests on init.

bplaat commented:
> This is cool! While playing around with it though, I had it freeze. Here's the stack trace: 

I have profiled the application some more and found out that most of the slowness comes from the PNG image decoding in particular the ZLib Decompressor.

Xexxa commented:
> Good point, I will work on a settings window where you can choose a map tiles provider or set a custom tiles API url. I will do this in another future pull request to keep this one small and reviewable.

The main reason I mention this was that we do not have one application(afaik) that is completely dependent on a third party service to function at all so it felt kinda weird to introduce that as a system application and not a port. If it was more generic(like being able to pick a provider) it would make more sense. Also want to make crystal clear that I'm not a maintainer, I do not know what(if any) policy we have on applications that require a single third party service to work and I do not mind if this PR gets merged. Just wanted to share my thinking ^^, also noted that @AtkinsSJ, who are a maintainer, left a review and didn't mention anything about this so I guess it's alright... ;-)

bplaat commented:
I have refactored and split out the commits to be more clear. And I have added a scale panel to the map widget in the bottom left corner inspired by [Leaflet.js](https://leafletjs.com/)

bplaat commented:
I forgot to add some code that shows an error message when you have no working internet connection. Otherwise you got just an empty looking gray map of unloaded images.

AtkinsSJ commented on Userland/Applications/Maps/main.cpp:
```diff
@@ -22,9 +23,9 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
-    options.center.latitude = 30.0;
-    options.center.longitude = 0.0;
-    options.zoom = 3;
+    options.center.latitude = Config::read_string("Maps"sv, "MapView"sv, "CenterLatitude"sv, "30"sv).to_double().release_value();
+    options.center.longitude = Config::read_string("Maps"sv, "MapView"sv, "CenterLongitude"sv, "0"sv).to_double().release_value();
+    options.zoom = Config::read_i32("Maps"sv, "MapView"sv, "Zoom"sv, 3);
```
You can't guarantee that `to_double()` will succeed, because it's reading from a user-editable config file. Rather than `release_value()`, use `value_or()` with the same default.

AtkinsSJ commented on Userland/Applications/Maps/main.cpp:
```diff
@@ -44,5 +45,11 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+
+    // Remember last map position
+    int exec = app->exec();
+    Config::write_string("Maps"sv, "MapView"sv, "CenterLatitude"sv, TRY(String::formatted("{}", maps->center().latitude)));
+    Config::write_string("Maps"sv, "MapView"sv, "CenterLongitude"sv, TRY(String::formatted("{}", maps->center().longitude)));
```
Use `String::number()` if you just want to convert a number to a string.

AtkinsSJ reviewed (changes requested):
Looking nice. Another comment I have that doesn't fit with any single part of the code though: It would look a little better if you gave the map a border. Easiest way to do that would be to wrap the MapWidget with a FrameWidget in the layout.

As a general style thing, we usually use `auto` a lot for declaring variables instead of actual type names, but I don't think it's a problem to be explicit with them.

bplaat commented on Userland/Applications/Maps/MapWidget.cpp:
```diff
@@ -0,0 +1,180 @@
+
+Optional<RefPtr<Gfx::Bitmap>> MapWidget::get_tile_image(int x, int y)
+{
+    // Get the right tile from tiles cache
+    for (Tile const& tile : m_tiles) {
+        if (tile.x == x && tile.y == y && tile.zoom == m_zoom) {
+            if (tile.image)
+                return tile.image;
+            return {};
+        }
+    }
```
The main reason I didn't use a hashmap was because I could not write a hash function for the x y, zoom key struct. There is only `pair_int_hash` in `AK/HashFunctions.h`. And I didn't want to write my own custom hash function.

bplaat commented:
I have fixed all your issues, it was a large rebase so I hope I didn't messed something up. I checked all the commits and they looked fine to me

gmta commented:
Remark on your commits: the last 5 commits all add to the first commit of adding the initial Maps application. I'd squash those so you'd get a single commit adding the application as a whole.

gmta commented on Userland/Applications/Maps/MapWidget.cpp:
```diff
@@ -0,0 +1,295 @@
+        set_zoom(m_zoom - 1);
+    } else {
+        set_zoom(m_zoom + 1);
+    }
```
I'd expect the mouse wheel zoom to actually zoom in to your mouse cursor (i.e. by panning the map to the cursor first)

bplaat replied:
You are right the code first only zoomed the center of the map. I have added the code that adjusts the zoom by the current mouse position.

gmta commented on Userland/Applications/Maps/MapWidget.h:
```diff
@@ -0,0 +1,121 @@
+    double const MAX_LATITUDE = 85.0511287798066;
+    size_t const MAX_TILES = 256;
+    float const PANEL_PADDING_X = 6;
+    float const PANEL_PADDING_Y = 4;
```
These can be `static constexpr`s

gmta commented on Userland/Applications/Maps/MapWidget.cpp:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    // Add tile when not in tiles
+    if (m_tiles.size() > MAX_TILES)
```
Shouldn't this be `>= MAX_TILES`?

bplaat commented on Userland/Applications/Maps/MapWidget.cpp:
```diff
@@ -0,0 +1,295 @@
+    m_attribution_height = PANEL_PADDING_Y + painter.font().pixel_size() + PANEL_PADDING_Y;
+    painter.fill_rect({ width() - m_attribution_width, height() - m_attribution_height, m_attribution_width, m_attribution_height }, Gfx::Color(255, 255, 255, 204));
+    Gfx::FloatRect attribution_text_rect { 0.0f, 0.0f, width() - PANEL_PADDING_X, height() - PANEL_PADDING_Y };
+    painter.draw_text(attribution_text_rect, m_attribution_text, Gfx::TextAlignment::BottomRight, Gfx::Color(51, 51, 51));
```
These colors are copied from [Leaflet.js](https://leafletjs.com/) and they match well with the currently used OSM map tile images. I would argue to not change these colors for now until we for example other themed map tiles. I see the drawed panels as a part of the complete map widget which uses an other style by default because the OSM tiles don't use the system theme colors.

gmta replied:
That makes sense, maybe add a comment about this so other contributors won't try to update it? It might help to extract the colors to separate variables as well.
